main/src/ca/uwaterloo/flix/language/phase/Monomorph.scala

```diff
@@ -394,6 +399,26 @@ object Monomorph {
         }
         Expression.Match(visitExp(exp, env0), rs, subst0(tpe), pur, eff, loc)
 
+      case Expression.MatchType(exp, rules, tpe, pur, eff, loc) =>
+        // make the tvars in `exp`'s type rigid
+        // so that Nil: List[x%123] can only match List[_]
+        val renv = exp.tpe.typeVars.foldLeft(RigidityEnv.empty) {
+          case (acc, Type.Var(sym, _)) => acc.markRigid(sym)
+        }
+        rules.collectFirst {
+          case MatchTypeRule(sym, t, body0)
+            // use the non-strict substitution
+            // to allow free type variables to match with anything
+            if Unification.unifiesWith(subst0.nonStrict(exp.tpe), subst0.nonStrict(t), renv) =>
+              // Generate a fresh symbol for the let-bound variable.
+              val freshSym = Symbol.freshVarSym(sym)
+              val env1 = env0 + (sym -> freshSym)
+              val e = visitExp(exp, env0)
+              val body = visitExp(body0, env1)
+              Expression.Let(freshSym, Modifiers.Empty, e, body, subst0(tpe), pur, eff, loc)
```

Comment:
It would be more precise/correct to use the purity/effect of the selected body (plus matchExp)?

Reply:
Yes, good point. That will help with optimization in later phases.